### PR TITLE
Stanza Converter

### PR DESCRIPTION
**Pull request by djaglowski**
**Description:** <Describe what has changed.>
This PR is a continuation of #1096, #1117, and #1124 (also see #1092 for additional context), and adds a concrete implementation for conversion from stanza log format to pdata.

Still not included are integration and testbed tests.

**Testing:**
Unit tests for converter.

### CODE REVIEW

File: receiver/stanzareceiver/converter.go
Comment (keitwb) [inline receiver/stanzareceiver/converter.go:161]: I think it does support arrays as of now.

File: receiver/stanzareceiver/converter.go
Comment (keitwb) [inline receiver/stanzareceiver/converter.go:193]: Are the pdata levels chosen arbitrary or is there some kind of logic to this?  That's fine if it is arbitrary I'm just curious.

Comment (keitwb) [reply receiver/stanzareceiver/converter.go:193]: Ok, I think I kind of see how the official levels are interspersed with the custom ones.

File: receiver/stanzareceiver/receiver.go
Comment (keitwb) [inline receiver/stanzareceiver/receiver.go:54]:
```suggestion
		r.logger.Info("Starting stanza receiver")
```

File: receiver/stanzareceiver/receiver.go
Comment (keitwb) [inline receiver/stanzareceiver/receiver.go:57]:
```suggestion
			err = fmt.Errorf("start stanza: %s", err)
```

Also should it bother starting the goroutine below if the Start method fails?

File: receiver/stanzareceiver/receiver.go
Comment (keitwb) [inline receiver/stanzareceiver/receiver.go:71]: This conversion routine is pretty inefficient in that there is an entire `pdata.Logs` created for each log record from Stanza.

Is it reasonable to expect multiple log records in close proximity with the same resource attributes?  If so you could pull off the channel and batch until the channel is empty and convert many at once into a single pdata.Logs (with potentially multiple pdata.ResourceLogs).  it isn't very easy to compare Resource since they are maps, and maybe doing that would overcome any performance advantages of batching.  But still something to consider.

File: receiver/stanzareceiver/converter.go
Comment (keitwb) [inline receiver/stanzareceiver/converter.go:40]: Would be nice to populate the InstrumentationLibrary object with the name and version of the stanza plugin used to generate the log.

File: receiver/stanzareceiver/converter.go
Comment (djaglowski) [reply receiver/stanzareceiver/converter.go]: Thanks for pointing this out. I missed the update. 

I've switched to using the new array functionality now.

File: receiver/stanzareceiver/converter.go
Comment (djaglowski) [reply receiver/stanzareceiver/converter.go:197]: I wouldn't say the conversion is arbitrary, but there's definitely some subjectivity to it.

Just trying to map from a 0-100 representation to a 0-24 representation, and it gets a little crowded around some levels.

File: receiver/stanzareceiver/converter.go
Comment (djaglowski) [reply receiver/stanzareceiver/converter.go:39]: Good call. Done.

One note - stanza currently doesn't make its version available in-code, except when built as a standalone executable. I considered using `runtime/debug.ReadBuildInfo()`, but this seems a bit messy. For now, I've hardcoded the version, but we'll look into updating our release process to make this programmatically available.

File: receiver/stanzareceiver/receiver.go
Comment (djaglowski) [reply receiver/stanzareceiver/receiver.go:72]: > Is it reasonable to expect multiple log records in close proximity with the same resource attributes?

Yes, quite often. I think you have a good point, and suspect that this is improvable.

I did actually take a stab at this earlier on, but did not see an improvement in the benchmarks at that time. I think it'll be worth taking a closer look though.

For now, I view the current benchmark results as acceptable, so it's up to you whether this is a necessary improvement at this point. My preference, for what it's worth, would be to leave batching out of the initial implementation and consider it a possible future optimization.

**Comment by djaglowski** [comment]
@keitwb I really appreciate the review! I've incorporated your feedback. Let me know what you think about the necessity of batching.

File: receiver/stanzareceiver/receiver.go
Comment (keitwb) [reply receiver/stanzareceiver/receiver.go:72]: I don't consider it essential for an initial version, and it is definitely not a trivial thing to do.

**Review by keitwb — APPROVED** [review]
LGTM.  @tigrannajaryan can you do a final review?

File: receiver/stanzareceiver/converter.go
Comment (keitwb) [reply receiver/stanzareceiver/converter.go:39]: Ok, yeah knowing it is stanza and having the version hardcoded for now is good enough.

**Comment by keitwb** [comment]
Oh also I just realized the other PRs are subsets of this one.  I think we can just consider this PR the main one and once it is merged the others should auto close.

**Comment by tigrannajaryan** [comment]
>I think we can just consider this PR the main one and once it is merged the others should auto close.

I would not want to do that. Let's keep the original PRs and commits and discussions. I would suggest to review in the order the PRs are submitted. If we merge this PR it will get squashed into one commit. History matters, let's preserve it.

**Comment by tigrannajaryan** [comment]
I believe https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1117 is the next one to review and merge (before this one).

**Comment by djaglowski** [comment]
Ok, so this is my understanding of where we're at with the three PRs:

1. We should process #1117 next, then #1124, then this one.
2. None of the feedback on this PR affected code in #1117, so I think that PR just needs a final review from @tigrannajaryan or another maintainer.
3. There were a few trivial changes made in response to feedback on this PR that affected lines submitted first in #1124. I am happy to backport those if that makes sense. Otherwise, they will be applied by this PR.

**Review by keitwb — APPROVED** [review]
Looks good.

**Review by keitwb — APPROVED** [review]

**Review by pmm-sumo — APPROVED** [review]

File: receiver/stanzareceiver/converter.go
Comment (djaglowski) [reply receiver/stanzareceiver/converter.go]: Thanks for the review @pmm-sumo. I've incorporated this suggestion.

**Review by bogdandrutu — APPROVED** [review]
This looks ready to be merged

**Pull request merged**
